File: rust/Python3/RustLexerBase.py
from typing import TextIO, Optional
from antlr4 import *
from antlr4.InputStream import InputStream

class RustLexerBase(Lexer):
    flt_mp = set()
    RustLexer = None
    def __init__(self, input: InputStream, output: TextIO = ...) -> None:
        super().__init__(input, output)
        self.token_lookbehind: tuple[Optional[Token], Optional[Token]] = (None, None)
        try:
            from .RustLexer import RustLexer
        except ImportError:
            from RustLexer import RustLexer
        RustLexerBase.RustLexer = RustLexer
        RustLexerBase.flt_mp = {
                RustLexer.STRING_LITERAL,
                RustLexer.RAW_STRING_LITERAL,
                RustLexer.BYTE_LITERAL,
                RustLexer.BYTE_STRING_LITERAL,
                RustLexer.RAW_BYTE_STRING_LITERAL,
                RustLexer.INTEGER_LITERAL,
                RustLexer.DEC_LITERAL,
                RustLexer.HEX_LITERAL,
                RustLexer.OCT_LITERAL,
                RustLexer.BIN_LITERAL,
                RustLexer.KW_SUPER,
                RustLexer.KW_SELFVALUE,
                RustLexer.KW_SELFTYPE,
                RustLexer.KW_CRATE,
                RustLexer.KW_DOLLARCRATE,
                RustLexer.RCURLYBRACE,
                RustLexer.RSQUAREBRACKET,
                RustLexer.RPAREN,
                RustLexer.KW_AWAIT,
                RustLexer.NON_KEYWORD_IDENTIFIER,
                RustLexer.RAW_IDENTIFIER,
                RustLexer.KW_MACRORULES,   
                RustLexer.GT 
            }

        """LOOK BEHIND TOKENS"""
    def nextToken(self):
        next: Token = super().nextToken()

        if next.channel == Token.DEFAULT_CHANNEL:
            self.token_lookbehind = self.token_lookbehind[1], next

        return next

    def SOF(self):
        return self._input.LA(-1) <= 0

    def next(self, expect) -> bool:
        if isinstance(expect, str):
            return chr(self._input.LA(1)) == expect
        else:
            return self._input.LA(1) == expect

     def nexti(self, expect) -> bool:
        if isinstance(expect, str):
            return chr(self._input.LA(1)) == expect
        else:
            return self._input.LA(1) == expect

    def floatDotPossible(self):
        next = chr(self._input.LA(1))
        # print(f'INFO: floatpossible ? {next} = {chr(next)}')
        # if isinstance(next, int):

        #     next = RustLexer.literalNames[next]
        #     next = next[1:-1]
        # if not isinstance(next, str):
        #     next = next.text

        if next in [".", "_"]:
            return False
        if next == "f":
            if chr(self._input.LA(2)) == "3" and chr(self._input.LA(3)) == "2":
                return True
            if chr(self._input.LA(2)) == "6" and self._input.LA(3) == "4":
                return True
            return False
        if next.isalpha():
            return False
        return True

    def floatLiteralPossible(self):
        prev, current = self.token_lookbehind
        
        if prev == None or current == None:
            return True
        elif current.type != RustLexerBase.RustLexer.DOT:
            return True
        else:
            return prev.type not in RustLexerBase.flt_mp 

File: rust/TypeScript/RustLexerBase.ts
import { CharStream, Token, Lexer } from 'antlr4';
import RustLexer from './RustLexer';

// Abstract class equivalent to RustLexerBase in Java
export default abstract class RustLexerBase extends Lexer {
    current: Token | null = null;
    previous: Token | null = null;

    constructor(input: CharStream) {
        super(input);
    }

    // Override the nextToken function to track the previous and current tokens
    nextToken(): Token {
        const next = super.nextToken();

        if (next.channel === Token.DEFAULT_CHANNEL) {
            this.previous = this.current;
            this.current = next;
        }

        return next;
    }

    SOF(): boolean {
        return this._input.LA(-1) <= 0;
    }

    // Check if the next character matches the expected character
    next(expect: string): boolean {
        return this._input.LA(1) === expect.charCodeAt(0);
    }
    nexti(expect: number): boolean {
        return this._input.LA(1) === input;
    }


    // Determine if a float dot is possible based on the next character
    floatDotPossible(): boolean {
        const next = this._input.LA(1);

        // only block . _ identifier after float
        if (next === '.'.charCodeAt(0) || next === '_'.charCodeAt(0)) {
            return false;
        }
        if (next === 'f'.charCodeAt(0)) {
            // 1.f32
            if (this._input.LA(2) === '3'.charCodeAt(0) && this._input.LA(3) === '2'.charCodeAt(0)) return true;
            // 1.f64
            if (this._input.LA(2) === '6'.charCodeAt(0) && this._input.LA(3) === '4'.charCodeAt(0)) return true;
            return false;
        }
        if ((next >= 'a'.charCodeAt(0) && next <= 'z'.charCodeAt(0)) ||
            (next >= 'A'.charCodeAt(0) && next <= 'Z'.charCodeAt(0))) {
            return false;
        }
        return true;
    }

    // Determine if a float literal is possible based on the previous tokens
    floatLiteralPossible(): boolean {
        if (this.current === null || this.previous === null) return true;
        if (this.current.type !== RustLexer.DOT) return true;

        switch (this.previous.type) {
            case RustLexer.CHAR_LITERAL:
            case RustLexer.STRING_LITERAL:
            case RustLexer.RAW_STRING_LITERAL:
            case RustLexer.BYTE_LITERAL:
            case RustLexer.BYTE_STRING_LITERAL:
            case RustLexer.RAW_BYTE_STRING_LITERAL:
            case RustLexer.INTEGER_LITERAL:
            case RustLexer.DEC_LITERAL:
            case RustLexer.HEX_LITERAL:
            case RustLexer.OCT_LITERAL:
            case RustLexer.BIN_LITERAL:
            case RustLexer.KW_SUPER:
            case RustLexer.KW_SELFVALUE:
            case RustLexer.KW_SELFTYPE:
            case RustLexer.KW_CRATE:
            case RustLexer.KW_DOLLARCRATE:
            case RustLexer.GT:
            case RustLexer.RCURLYBRACE:
            case RustLexer.RSQUAREBRACKET:
            case RustLexer.RPAREN:
            case RustLexer.KW_AWAIT:
            case RustLexer.NON_KEYWORD_IDENTIFIER:
            case RustLexer.RAW_IDENTIFIER:
            case RustLexer.KW_MACRORULES:
                return false;
            default:
                return true;
        }
    }
}
